src/Compiler/Utilities/TypeHashing.fs

```diff
@@ -357,68 +358,112 @@ module HashTastMemberOrVals =
-            tcref1.Stamp = tcref2.Stamp
-            && nullnessEquals n1 n2
-            && typeInstStampsEqual tinst1 tinst2
-        | TType_anon(info1, tys1), TType_anon(info2, tys2) -> info1.Stamp = info2.Stamp && typeInstStampsEqual tys1 tys2
-        | TType_tuple(c1, tys1), TType_tuple(c2, tys2) -> c1 = c2 && typeInstStampsEqual tys1 tys2
-        | TType_forall(tps1, tau1), TType_forall(tps2, tau2) -> stampEquals tau1 tau2 && typarsStampsEqual tps1 tps2
-        | TType_var(r1, n1), TType_var(r2, n2) -> r1.Stamp = r2.Stamp && nullnessEquals n1 n2
-        | TType_measure m1, TType_measure m2 -> measureStampEquals m1 m2
-        | _ -> false
-
-    let inline hashStamp (x: Stamp) : Hash = uint x * 2654435761u |> int
-
-    // The idea is to keep the illusion of immutability of TType.
-    // This hash must be stable during compilation, otherwise we won't be able to find keys or evict from the cache.
-    let rec hashTType ty : Hash =
+/// <summary>
+/// StructuralUtilities: produce a conservative structural fingerprint of <c>TType</c>.
+///
+/// Current (sole) usage:
+///   Key in the typeSubsumptionCache. The key must never give a false positive
+///   (two non-subsuming types producing identical token sequences). False negatives
+///   are acceptable and simply reduce cache hit rate.
+///
+/// Properties:
+///   * Uses per-compilation stamps (entities, typars, anon records, measures).
+///   * Emits shape for union cases (declaring type stamp + case name), tuple structness,
+///     function arrows, forall binders, nullness, measures, generic arguments.
+///   * Unknown/variable nullness => NeverEqual token to force inequality (avoid unsound hits).
+///
+/// Non-goals:
+///   * Cross-compilation stability.
+///   * Perfect canonicalisation or alpha-equivalence collapsing.
+///
+/// </summary>
+module StructuralUtilities =
+    [<Struct; CustomEquality; NoComparison>]
+    type NeverEqual =
+        struct
+            interface System.IEquatable<NeverEqual> with
+                member _.Equals _ = false
+
+            override _.Equals _ = false
+            override _.GetHashCode() = 0
+        end
+
+        static member Singleton = NeverEqual()
+
+    [<Struct; NoComparison; RequireQualifiedAccess>]
+    type TypeToken =
+        | Stamp of stamp: Stamp
+        | UCase of name: string
+        | Nullness of nullness: NullnessInfo
+        | TupInfo of b: bool
+        | MeasureOne
+        | MeasureRational of rational: Rational
+        | NeverEqual of never: NeverEqual
+
+    type TypeStructure = TypeToken[]
+
+    [<Literal>]
+    let private initialTokenCapacity = 4
+
+    let inline toNullnessToken (n: Nullness) =
+        match n with
+        | Nullness.Known k -> TypeToken.Nullness k
+        // If nullness is not known we must treat the types as not equal for caching purposes.
+        | Nullness.Variable _ -> TypeToken.NeverEqual NeverEqual.Singleton
```

Comment:
I would rather do a `TryEvaluate` here.
A variable, but solved, nullness should be treated the same as a known one.


The only problematic area would be unsolved variable nullness